aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java

```diff
@@ -152,35 +212,130 @@ private static void setEgressOperation(SpanData span, AttributesBuilder builder)
    * and RPC attributes to use here, but this is a sufficient starting point.
    */
   private static void setRemoteServiceAndOperation(SpanData span, AttributesBuilder builder) {
+    String remoteService;
+    String remoteOperation;
     if (isKeyPresent(span, AWS_REMOTE_SERVICE) || isKeyPresent(span, AWS_REMOTE_OPERATION)) {
-      setRemoteService(span, builder, AWS_REMOTE_SERVICE);
-      setRemoteOperation(span, builder, AWS_REMOTE_OPERATION);
+      remoteService = getRemoteService(span, AWS_REMOTE_SERVICE);
+      remoteOperation = getRemoteOperation(span, AWS_REMOTE_OPERATION);
     } else if (isKeyPresent(span, RPC_SERVICE) || isKeyPresent(span, RPC_METHOD)) {
-      setRemoteService(span, builder, RPC_SERVICE);
-      setRemoteOperation(span, builder, RPC_METHOD);
+      remoteService = getRemoteService(span, RPC_SERVICE);
+      remoteOperation = getRemoteOperation(span, RPC_METHOD);
     } else if (isKeyPresent(span, DB_SYSTEM) || isKeyPresent(span, DB_OPERATION)) {
-      setRemoteService(span, builder, DB_SYSTEM);
-      setRemoteOperation(span, builder, DB_OPERATION);
-    } else if (isKeyPresent(span, FAAS_INVOKED_PROVIDER) || isKeyPresent(span, FAAS_INVOKED_NAME)) {
-      setRemoteService(span, builder, FAAS_INVOKED_PROVIDER);
-      setRemoteOperation(span, builder, FAAS_INVOKED_NAME);
+      remoteService = getRemoteService(span, DB_SYSTEM);
+      remoteOperation = getRemoteOperation(span, DB_OPERATION);
+    } else if (isKeyPresent(span, FAAS_INVOKED_NAME) || isKeyPresent(span, FAAS_TRIGGER)) {
+      remoteService = getRemoteService(span, FAAS_INVOKED_NAME);
+      remoteOperation = getRemoteOperation(span, FAAS_TRIGGER);
     } else if (isKeyPresent(span, MESSAGING_SYSTEM) || isKeyPresent(span, MESSAGING_OPERATION)) {
-      setRemoteService(span, builder, MESSAGING_SYSTEM);
-      setRemoteOperation(span, builder, MESSAGING_OPERATION);
+      remoteService = getRemoteService(span, MESSAGING_SYSTEM);
+      remoteOperation = getRemoteOperation(span, MESSAGING_OPERATION);
     } else if (isKeyPresent(span, GRAPHQL_OPERATION_TYPE)) {
-      builder.put(AWS_REMOTE_SERVICE, GRAPHQL);
-      setRemoteOperation(span, builder, GRAPHQL_OPERATION_TYPE);
+      remoteService = GRAPHQL;
+      remoteOperation = getRemoteOperation(span, GRAPHQL_OPERATION_TYPE);
     } else {
-      logUnknownAttribute(AWS_REMOTE_SERVICE, span);
-      builder.put(AWS_REMOTE_SERVICE, UNKNOWN_REMOTE_SERVICE);
-      logUnknownAttribute(AWS_REMOTE_OPERATION, span);
-      builder.put(AWS_REMOTE_OPERATION, UNKNOWN_REMOTE_OPERATION);
+      // try to derive RemoteService and RemoteOperation from the other un-directive attributes
+      remoteService = generateRemoteService(span);
```

Comment:
It's hard to tell the difference between `generateRemoteService/Operation` and `getRemoteService/Operation`, only that `generate*` does some specific to HTTP/NET spans which makes me feel the method name is ambiguous.

I'm thinking if it will be better to assign `UNKNOWN*` as the initial values to `remoteService` and `remoteOperation`, and only do HTTP/NET handling using function like `getRemote*FromNETSpan` in the `else` block.

Reply:
Thanks! It's much clear now.